### PR TITLE
Set response headers based on Rack version

### PR DESCRIPTION
**Pull request by schinery**
https://github.com/ruby-grape/grape/issues/2354

### CODE REVIEW

**Comment by dblock** [comment]
I like this better than https://github.com/ruby-grape/grape/pull/2356. Same question as in that issue, let's dig up why this was changed.

**Comment by schinery** [comment]
> One option would be to major-version increment Grape as part of this PR and do the same as in Rack

Do you need something adding to this PR for this? 

I also pushed fixing the integration spec paths and updated changelog.

**Comment by dblock** [comment]
Let's see if other maintainers have an opinion. If we major increment the version here we can make the backwards incompatible change. I am 50/50 on what's best.

**Comment by dblock** [comment]
@ioquatix would like your eyes and opinion here if you have a moment?

**Comment by dblock** [comment]
@schinery looks like we may need to do extra work to support Rack 1.x.

**Comment by schinery** [comment]
> @schinery looks like we may need to do extra work to support Rack 1.x.

Ah, seems that `Rack::RELEASE` was added after 1.x., but there is `Rack.release` in all 3 versions ([1.x.](https://github.com/rack/rack/blob/1-6-stable/lib/rack.rb#L22), [2.x.](https://github.com/rack/rack/blob/2-2-stable/lib/rack/version.rb#L26) and [3.x.](https://github.com/rack/rack/blob/3-0-stable/lib/rack/version.rb#L31)) so switched to use that for the version check.

**Comment by schinery** [comment]
> @schinery Let's increment the version to 1.9 as part of this PR since at the very least it's a big change, and add a section to UPGRADING?

@dblock like this? https://github.com/ruby-grape/grape/pull/2355/commits/8c109338224a182043074f5406f7ef18b2f95631

File: CHANGELOG.md
Comment (dblock) [inline CHANGELOG.md:11]: We didn't release 1.9.0 yet, so remove this and say above `### 1.9.0 (Next)`.

File: README.md
Comment (dblock) [inline README.md:165]: Current stable is 1.8.0, revert this file change.

File: UPGRADING.md
Comment (dblock) [inline UPGRADING.md:17]: Add a link for more information see this PR like below.

I think we need more text explaining what happens in what code. Give an example of an API that sets one of these headers, or explain when they are set?

File: UPGRADING.md
Comment (dblock) [inline UPGRADING.md:8]: This should say something like "As per https://github.com/rack/rack/issues/1592 Rack 3.0 is enoforcing the HTTP/2 semantics, and thus treats all headers as lowercase. Starting with Grape 1.9.0, the following headers are now lowercase: ..."

**Comment by schinery** [comment]
@dblock hopefully done all the PR feedback correctly, with the exception of "I think we need more text explaining what happens in what code. Give an example of an API that sets one of these headers, or explain when they are set?" as not 100% sure what you want for this.

Do you have an example in mind?

**Comment by schinery** [comment]
Also, just going through the README, and wondering if the following needs updating:

* https://github.com/ruby-grape/grape#headers
* https://github.com/ruby-grape/grape#rails-3x

There is also https://github.com/ruby-grape/grape#header-case-handling, which this PR hasn’t altered as it has only changed the two hardcoded headers, but should there be a code change to lower case these for Rack 3 too?

**Comment by dblock** [comment]
@schinery Yes to all your questions above about https://github.com/ruby-grape/grape/issues/2254 and related issues. I can merge as is, or do you think we alter all related behavior for Rack 3 (i.e. resolve #2254) in this one PR or separately?

**Comment by schinery** [comment]
@dblock just trying to get a sense of the other changes needed to resolve #2254 to see if we do it in this PR, so far I have:

1: Update `#transform_header` in https://github.com/ruby-grape/grape/blob/master/lib/grape/request.rb#L49 to be Rack specific, so capitalize for Rack < 2, lowercase for Rack >= 3. Think this is fairly straightforward as long as we're ok to use the same Rack check used in the headers class?

2: Fix the specs, like https://github.com/ruby-grape/grape/blob/master/spec/grape/request_spec.rb#L121, so they are Rack version specific. Am guessing this would mean that they would go in the integration specs like we've added in this PR?

3: Update the README, such as https://github.com/ruby-grape/grape#header-case-handling, so that it has changes for both header casings.

4: Don't think we need to do anything for the `#header` method defined in https://github.com/ruby-grape/grape/blob/master/lib/grape/dsl/headers.rb#L11, as this is down to whatever people use for the header key.

Not sure what else, partly as this is my first time truly digging around the code. Anything else you think is needed?

**Comment by schinery** [comment]
Does also make me wonder... would it be an idea to leave version 1 of Grape supporting Rack < 3 and create a version 2 of Grape that supports Rack >= 3. 

There would be more overhead for maintaining etc depending how long you want to maintain version 1, but the code would be cleaner without the Rack version detection code in it. 

Happy to continue with https://github.com/ruby-grape/grape/pull/2355/#issuecomment-1773821081 though, let me know your thoughts.

**Comment by schinery** [comment]
@dblock more for https://github.com/ruby-grape/grape/pull/2355#issuecomment-1773821081

I threw [this PR](https://github.com/schinery/grape/pull/1) together quick just to look at what might need changing (minus the README changes), especially which specs would need updating. There were 5 specs that need the header in the right case based on what Rack version is used in the test run.

**Comment by dblock** [comment]
> Does also make me wonder... would it be an idea to leave version 1 of Grape (1.x branch off of master) supporting Rack < 3 and create a version 2 of Grape (current master) that supports Rack >= 3.

I would be totally fine with incrementing major to 2.0 as part of this PR, and lock it at Rack >= 3, instead of trying to do both if you prefer?

I don't love the idea of also maintaining a 1.x. Are you interested in doing so? It's just a 1.x branch and releasing.

**Comment by schinery** [comment]
>I would be totally fine with incrementing major to 2.0 as part of this PR, and lock it at Rack >= 3, instead of trying to do both if you prefer?

It just feels a bit cleaner. As you saw in https://github.com/schinery/grape/pull/1 (which again was more to highlight than a solution) some of the issues are in the testing, plus I guess you also have to have update the docs to support both Rack versions. 

> I don't love the idea of also maintaining a 1.x. Are you interested in doing so? It's just a 1.x branch and releasing.

Not really interested either, but I guess then you have something if you need to patch 1.x. in the future, and I guess then you could then just release a 1.8.1 or 1.9.0 release with https://github.com/ruby-grape/grape/pull/2360 and updated README etc, with any new features etc going in the 2.x version.

If it is decided that this is approach to take then is this what needs to happen?

* Update this PR to:
  * Bump Grape to 2.0.0
  * Lock Rack to >= 3
  * Potentially lock ActiveSupport to >= 7.1 (not sure if 100% needed but Rails 7.0 uses Rack 2.0)
  * Update the test matrix to `gemfile: [rack_3_0, rails_7_1]` only and clean out old gemfiles
  * Update the `transform_headers` to downcase only and update the specs where needed
  * Update the README, CHANGELOG and UPGRADING docs
* You'd create a 1.x branch and then we'd need a PR for that branch that:
  * Bumped Grape to 1.8.1 or 1.9.0
  * Lock Rack < 3
  * Potentially lock ActiveSupport to < 7.1
  * Update the test matrix to not include `rack_3_0` and  `rails_7_1`
   * Update the README, CHANGELOG and UPGRADING docs 

Does it feel like we need a second opinion on this too?

**Comment by schinery** [comment]
One thing I was wrong about... it wasn't Rails 7.1 that bumped Rack to 3, as the deps in the Gemfile.lock are as follows:

```Gemfile
    actionpack (7.1.1)
      actionview (= 7.1.1)
      activesupport (= 7.1.1)
      nokogiri (>= 1.8.5)
      rack (>= 2.2.4)
      rack-session (>= 1.0.1)
      rack-test (>= 0.6.3)
      rails-dom-testing (~> 2.2)
      rails-html-sanitizer (~> 1.6)
```

However, I've just been playing about with locking to Rack 3 in a branch and I get this error using `rails_7_0.gemfile`. 

```
And because actionpack >= 7.0.5, < 7.1.0 depends on rack >= 2.2.4, < 3.A
  and every version of grape depends on rack >= 3.0.0,
  every version of grape is incompatible with rails >= 7.0.0, < 7.1.0.beta1.
So, because rails_7_0.gemfile depends on rails ~> 7.0.0
  and rails_7_0.gemfile depends on grape >= 0,
  version solving has failed.
```

So, if Grape was locked to Rack 3 it looks like potentially that version might only work with Rails 7.1. 

Which means I'm now flip flopping between whether Rack should be locked in Grape or not. Suspect not...

**Comment by schinery** [comment]
I've updated https://github.com/schinery/grape/pull/1 with the additional changes I think are needed to support both Rack versions in Grape. I haven't addressed your additional comments in that PR, partly as my brain is tied in knots.

**Comment by schinery** [comment]
Merged https://github.com/schinery/grape/pull/1 in. 

Will ponder the "a case-insensitive structure" for the request headers, would be ace if there was a way to do it but there is nothing screaming out at me at the moment.

**Comment by dblock** [comment]
> Will ponder the "a case-insensitive structure" for the request headers, would be ace if there was a way to do it but there is nothing screaming out at me at the moment.

In https://github.com/schinery/grape/blob/51b081cef9d4d91cb75f80dd33222805612c7b95/lib/grape/request.rb#L38

This builds a `Hash`. You could try with `ActiveSupport::HashWithIndifferentAccess`, I don't remember if it's also case-insensitive.

File: UPGRADING.md
Comment (dblock) [reply UPGRADING.md:27]: I still want an example here that shows what happens when I was doing `request.heaeders['Content-Type']` before and what I should be doing now. Should I change that code to `request.headers[Grape::Http::Headers::CONTENT_TYPE]` for example? What will the server return before/after?

File: lib/grape/request.rb
Comment (dblock) [inline lib/grape/request.rb:50]: This executes in a tight loop, which is bad for performance. You want this check outside. We also have it in a few places. It should look like this:

```ruby
if Grape::Rack.rack3?
  def transform_header(header)
      -header[5..].split('_').map(&:capitalize).join('-')
  end
else
  def transform_header(header)
      -header[5..].tr('_', '-').downcase
  end
end
```

File: spec/grape/api_spec.rb
Comment (dblock) [inline spec/grape/api_spec.rb:692]: Can we get rid of `rack_versioned_headers` and instead do `Grape::HTTP::Headers::CONTENT_TYPE`?

File: spec/support/headers_helpers.rb
Comment (dblock) [inline spec/support/headers_helpers.rb:6]: Get rid of this class, use `Grape::HTTP::Headers::CACHE_CONTROL` and friends.

**Review by dblock — CHANGES_REQUESTED** [review]
Getting there ...

File: lib/grape/request.rb
Comment (schinery) [reply lib/grape/request.rb]: Added `.rack3?` method and done this change, although I've added it to `Grape` rather than `Grape::Rack` otherwise there was a load of Rack errors raised because we would need to change all the Rack references to `::Rack`.

Can do this though if really needs to be `Grape::Rack`

File: spec/support/headers_helpers.rb
Comment (schinery) [reply spec/support/headers_helpers.rb]: The only issue with this is what to do in the specs that are testing with non `Grape::HTTP::Headers` headers, such as `Grape-Likes-Symbolic` or `X-Access-Token`, as these need to be Rack versioned in the specs. 

I've removed the headers helper though and put these inline, so they are defined in whatever specs needed them

**Comment by schinery** [comment]
> > Will ponder the "a case-insensitive structure" for the request headers, would be ace if there was a way to do it but there is nothing screaming out at me at the moment.
> 
> In https://github.com/schinery/grape/blob/51b081cef9d4d91cb75f80dd33222805612c7b95/lib/grape/request.rb#L38
> 
> This builds a `Hash`. You could try with `ActiveSupport::HashWithIndifferentAccess`, I don't remember if it's also case-insensitive.

Yeah, it doesn't...

```sh
irb(main):001:0> hash = ActiveSupport::HashWithIndifferentAccess.new(a: 1)
=> {"a"=>1}
irb(main):002:0> hash["A"]
=> nil
irb(main):003:0>
```

Whether something like [this](https://stackoverflow.com/a/2030565/1517515) would work. I'm not sure what code changes would be required though.

File: spec/grape/api/custom_validations_spec.rb
Comment (dblock) [inline spec/grape/api/custom_validations_spec.rb:168]: Could use `x_access_token_header` here.

**Review by dblock — APPROVED** [review]
I'm going to merge and let it sit.

File: spec/grape/api/custom_validations_spec.rb
Comment (schinery) [reply spec/grape/api/custom_validations_spec.rb:168]: Yeah, tried that, but got...

```
     NameError:
       undefined local variable or method `x_access_token_header' for #<#<Class:0x0000ffff8f8f9b18>:0x0000ffff928d97e8 @attrs=[:admin_field], @option=true, @required=false, @scope=#<Grape::Validations::ParamsScope:0x0000ffff8f8f96b8 @element=nil, @element_renamed=nil, @parent=nil, @api=#<Class:0x0000ffff8f8f9898>, @optional=false, @type=Hash, @group=nil, @dependent_on=nil, @declared_params=nil, @index=nil>, @fail_fast=false, @allow_blank=nil>
```

which was annoying, as don't like the defining it twice either.

File: UPGRADING.md
Comment (ioquatix) [inline UPGRADING.md:8]: > Rack 3.0 is enforcing the HTTP/2 semantics, and thus treats all headers as lowercase. Starting with Grape 1.9.0, headers will be cased based on what version of Rack you are using.

Rack 3 is following the HTTP/2+ semantics which require header names to be lower case. To avoid compatibility issues, starting with Grape 1.9.0, headers will be cased based on what version of Rack you are using.

File: lib/grape.rb
Comment (ioquatix) [inline lib/grape.rb:46]: If you are just trying to detect whether to use lower case headers or not, I suggest a slightly different approach:

```ruby
LOWER_CASE_HEADERS = Rack::CONTENT_TYPE == "content-type"
```

You could then use that directly or in a method.

File: lib/grape/dsl/inside_route.rb
Comment (ioquatix) [inline lib/grape/dsl/inside_route.rb:331]: You can use `Rack::CONTENT_LENGTH` if that suits your use case.

**Review by ioquatix — COMMENTED** [review]
This generally looks okay to me.

The biggest issue is, people trying to support both Rack 2 and Rack 3 will need to be careful about the assertions. Direct assertions against the hash can be problematic. It's important to have a proper normalization step in assertions IMHO, as the expectation should be "A header named X has a value Y", but the case of X can depend on the HTTP version, server, etc. In Rails, there were many instances of header key and value case which had ossified to whatever Rack (the gem) was generating, even thought that was only a subset of valid possible results.

I know this was a pretty annoying change in some regards, but overall it will simplify things and aligns the code with the HTTP specification more closely in a way which makes it easier to know that code is doing the right thing. e.g. https://github.com/rack/rack-test/issues/229

**Comment by dblock** [comment]
Thanks @ioquatix ! @schinery will you extract the comments into another PR please?

**Comment by schinery** [comment]
https://github.com/ruby-grape/grape/pull/2362

**Pull request merged**
